pkgs/applications/misc/dmenu-rs/default.nix

```
@@ -0,0 +1,98 @@
+{ stdenv
+, rustPlatform
+, lib
+, fetchFromGitHub
+, cargo
+, clang
+, expat
+, fontconfig
+, libXft
+, libXinerama
+, m4
+, pkg-config
+, python3
+}:
+
+# The dmenu-rs package has extensive plugin support. However, this derivation
+# only builds the package with the default set of plugins. If you'd like to
+# further customize dmenu-rs you can build it from the source.
+# See: https://github.com/Shizcow/dmenu-rs#plugins
+stdenv.mkDerivation rec {
+  pname = "dmenu-rs";
+  version = "5.5.1";
+
+  src = fetchFromGitHub {
+    owner = "Shizcow";
+    repo = pname;
+    rev = version;
+    sha256 = "sha256-WpDqBjIZ5ESnoRtWZmvm+gNTLKqxL4IibRVCj0yRIFM=";
+  };
+
+  nativeBuildInputs = [
+    cargo
+    clang
+    m4
+    pkg-config
+    python3
+    rustPlatform.bindgenHook
+    rustPlatform.cargoSetupHook
+  ];
+
+  buildInputs = [
+    expat
+    fontconfig
+    libXft
+    libXinerama
+  ];
+
+  # The dmenu-rs repository does not include a Cargo.lock because of its
+  # dynamic build and plugin support. Generating it with make and checking it
+  # in to nixpkgs here was the easiest way to supply it to rustPlatform.
+  # See: https://github.com/Shizcow/dmenu-rs/issues/34#issuecomment-757415584
+  cargoDeps = rustPlatform.importCargoLock {
+    lockFile = ./Cargo.lock;
+  };
+
+  # The rust-xcb dependency dynamically generates rust code at build time.
+  # This derivation uses nixpkgs rust functions that vendor each cargo
+  # dependency's source code into the READ-ONLY nix store. To avoid the code
+  # generation step failing, we copy the rust-xcb source out of the nix store
+  # and make it writeable. Also, we remove the build's hardcoded c compiler.
+  # See: https://github.com/rust-x-bindings/rust-xcb/tree/v0.8.2
+  patchPhase = ''
+    runHook prePatch
+
+    substituteInPlace config.mk --replace "clang" ""
+
+    chmod +w "$NIX_BUILD_TOP/cargo-vendor-dir"
+    mkdir -p "$NIX_BUILD_TOP/cargo-vendor-dir/xcb-0.8.2-readwrite"
+    cp -r --no-preserve=mod "$NIX_BUILD_TOP/cargo-vendor-dir/xcb-0.8.2/." "$NIX_BUILD_TOP/cargo-vendor-dir/xcb-0.8.2-readwrite"
+    unlink "$NIX_BUILD_TOP/cargo-vendor-dir/xcb-0.8.2"
+    mv "$NIX_BUILD_TOP/cargo-vendor-dir/xcb-0.8.2-readwrite" "$NIX_BUILD_TOP/cargo-vendor-dir/xcb-0.8.2"
+
+    runHook postPatch
+  '';
+
+  postPatch = ''
+    cp ${./Cargo.lock} src/Cargo.lock
+  '';
+
+  cargoRoot = "src";
+
+  preConfigure = ''
+    sed -i "s@PREFIX = /usr/local@PREFIX = $out@g" config.mk
+  '';
```

Some more corners to 

1. this should be a postPatch
2. It is better to remove this setting from config.mk:

```suggestion
  postPatch = ''
    sed -i "s@PREFIX = /usr/local@@g" config.mk
  '';
```
3. Use `installFlags = [ "PREFIX=$(out)" ];`.  (Indeed, maybe the previous step becomes unneeded if this flag works.)